packages/nitro-protocol/docs/app-devs/recommended-patterns.md

```diff
@@ -0,0 +1,284 @@
+
+function toRPSData(appData: AppData): RPSData {
+  let positionType;
+  switch (appData.type) {
+    case 'start':
+      positionType = PositionType.Start;
+      break;
+    case 'roundProposed':
+      positionType = PositionType.RoundProposed;
+      break;
+    case 'roundAccepted':
+      positionType = PositionType.RoundAccepted;
+      break;
+    case 'reveal':
+      positionType = PositionType.Reveal;
+      break;
+  }
+  const defaults: RPSData = {
+    positionType,
+    stake: bigNumberify(0).toString(),
+    preCommit: HashZero,
+    playerAWeapon: Weapon.Rock,
+    playerBWeapon: Weapon.Rock,
+    salt: randomHex(64),
+  };
+
+  return {...defaults, ...appData};
+}
+
+export function encodeAppData(appData: AppData): string {
+  return encodeRPSData(toRPSData(appData));
+}
+
+export function encodeRPSData(rpsData: RPSData): string {
+  return defaultAbiCoder.encode(
+    [
+      'tuple(uint8 positionType, uint256 stake, bytes32 preCommit, uint8 playerAWeapon, uint8 playerBWeapon, bytes32 salt)',
+    ],
+    [rpsData]
+  );
+}
+
+export function decodeAppData(appDataBytes: string): AppData {
+  const parameters = defaultAbiCoder.decode(
+    [
+      'tuple(uint8 positionType, uint256 stake, bytes32 preCommit, uint8 playerAWeapon, uint8 playerBWeapon, bytes32 salt)',
+    ],
+    appDataBytes
+  )[0];
+
+  const positionType = parameters[0] as PositionType;
+  const stake = parameters[1].toString();
+  const preCommit = parameters[2];
+  const playerAWeapon = parameters[3];
+  const playerBWeapon = parameters[4];
+  const salt = parameters[5];
+
+  switch (positionType) {
+    case PositionType.Start:
+      const start: Start = {
+        type: 'start',
+        stake,
+      };
+      return start;
+    case PositionType.RoundProposed:
+      const roundProposed: RoundProposed = {
+        type: 'roundProposed',
+        stake,
+        preCommit,
+      };
+      return roundProposed;
+    case PositionType.RoundAccepted:
+      const roundAccepted: RoundAccepted = {
+        type: 'roundAccepted',
+        stake,
+        preCommit,
+        playerBWeapon,
+      };
+      return roundAccepted;
+    case PositionType.Reveal:
+      const reveal: Reveal = {
+        type: 'reveal',
+        playerAWeapon,
+        playerBWeapon,
+        salt,
+        stake,
+      };
+      return reveal;
+    default:
+      return unreachable(positionType);
+  }
+}
+
+export function hashPreCommit(weapon: Weapon, salt: string) {
+  return keccak256(defaultAbiCoder.encode(['uint256', 'bytes32'], [weapon, salt]));
+}
+```
+
+## Wrapping the channelClient
+
+We recommend that you wrap the channelClient in an `AppSecificChannelClient`, which can leverage your app-specific coding [as above](#application-coding). This additional layer of abstraction should result in cleaner code for the rest of your application.
+
+Here's a sketch of what we did with Rock Paper Scissors:
+
+```typescript
+export class RPSChannelClient {
```

Comment:
This doesn't look like it has all the functionality?

Reply:
No, I stripped most of it out. There's already a lot of code on this page, and I didn't want to overdo it. I also wanted to reflect the fact that we don't (yet) have a very rigid or fixed set of patterns for these wrapping classes. Hence "sketch".

Reply:
Have added a placeholder to make it clearer that the example is incomplete, as well as a GH link to the full thing.